README

 mad - MPEG audio decoder
 Copyright (C) 2000-2001 Robert Leslie

 $Id: README,v 1.46 2001/10/18 08:34:41 rob Exp $

===============================================================================

INTRODUCTION

  MAD is a high-quality MPEG audio decoder. It currently supports MPEG-1 and
  the MPEG-2 extension to Lower Sampling Frequencies, as well as the
  so-called MPEG 2.5 format. All three audio layers (Layer I, Layer II, and
  Layer III a.k.a. MP3) are fully implemented.

  MAD does not yet support MPEG-2 multichannel audio (although it should be
  backward compatible with such streams) nor does it currently support AAC.

  MAD has the following special features:

    - 24-bit PCM output
    - 100% fixed-point (integer) computation
    - completely new implementation based on the ISO/IEC standards
    - distributed under the terms of the GNU General Public License (GPL)

  Because MAD provides full 24-bit PCM output, applications using MAD are
  able to produce high quality audio. Even when the output device supports
  only 16-bit PCM, applications can use the extra resolution to increase the
  audible dynamic range through the use of dithering or noise shaping.

  Because MAD uses integer computation rather than floating point, it is
  well suited for architectures without a floating point unit. All
  calculations are performed with a 32-bit fixed-point integer
  representation.

  Because MAD is a new implementation of the ISO/IEC standards, it is
  unencumbered by the errors and copyrights of other implementations. MAD is
  NOT a derivation of the ISO reference source or any other code.
  Considerable effort has been expended to ensure a correct implementation,
  even in cases where the standards are ambiguous or misleading.

  Because MAD is distributed under the terms of the GPL, its redistribution
  is not generally restricted, so long as the terms of the GPL are followed.
  This means MAD can be incorporated into other software as long as that
  software is also distributed under the GPL. (Should this be undesirable,
  alternate arrangements may be possible by contacting the author.)

===============================================================================

ABOUT THE CODE

  The code currently consists of an MPEG audio decoding library (libmad) and
  a simple front-end called `madplay' that supports several audio output
  modules:

    - an Open Sound System interface module (for Linux, et al.)
    - a Sun audio interface module (for Solaris, NetBSD, et al.)
    - a Win32 audio interface module (for Windows 95/98/NT, et al.)
    - a raw PCM output module
    - a WAVE file output module (*.wav)
    - a Sun/NeXT audio file output module (*.au, *.snd)
    - a hex output module (for debugging and compliance testing)
    - a null module (for timing the decoder)

  The code is optimized and performs very well, although specific
  improvements can still be made. The output from the decoder library
  consists of 32-bit signed linear fixed-point values which can be easily
  scaled for any size PCM output, up to 24 bits per sample.

  The API for libmad is likely to change, so it is not yet documented.
  However, the header file `libmad/mad.h' may be informative, as may also be
  the sample code in `minimad.c'. Note that the `libmad/mad.h' file is
  automatically generated, and will not exist until you have built the
  library.

  There are two APIs available, one high-level, and the other low-level.
  With the low-level API, each step of the decoding process must be handled
  explicitly, offering the greatest amount of control. With the high-level
  API, after callbacks are configured, a single routine will decode an
  entire bitstream.

  The high-level API may either be used synchronously or asynchronously. If
  used asynchronously, decoding will occur in a separate process.
  Communication is possible with the decoding process by passing control
  messages.

  The `madplay' front-end is written to use the high-level synchronous API.

  The file `minimad.c' contains an example usage of the libmad API which is
  less complex than `madplay.c' and shows only the bare minimum required to
  implement a useful decoder. It expects a regular file to be redirected to
  standard input, and it sends decoded 16-bit signed little-endian PCM
  samples to standard output. It will abort at the first discovery of a
  decoding error. Note that the scale() routine in this code is only
  provided as an example; it rounds MAD's high-resolution samples down to 16
  bits, but does not perform any dithering or noise shaping. It is therefore
  not recommended to use this routine as-is in your own code if sound
  quality is important.

Integer Performance

  To get the best possible performance, it is recommended that an assembly
  version of the fixed-point multiply and related routines be selected.
  Several such assembly routines have been written for various CPUs.

  If an assembly version is not available, a fast approximation version will
  be used. This will result in reduced accuracy of the decoder.

  Alternatively, if 64-bit integers are supported as a datatype by the
  compiler, another version can be used which is much more accurate.
  However, using an assembly version is generally much faster and just as
  accurate.

  More information can be gathered from the `libmad/fixed.h' header file.

  MAD's CPU-intensive subband synthesis routine can be further optimized at
  the expense of a slight loss in output accuracy due to a modified method
  for fixed-point multiplication with a small windowing constant. While this
  is helpful for performance and the output accuracy loss is generally
  undetectable, it is disabled by default and must be explicitly enabled.

  Under some architectures, other special optimizations may also be
  available.

Audio Quality

  The output from MAD has been tested and found to satisfy the ISO/IEC
  11172-4 computational accuracy requirements for compliance. In most
  configurations, MAD is a Full Layer III ISO/IEC 11172-3 audio decoder as
  defined by the standard.

  When the approximation version of the fixed-point multiply is used, MAD is
  a limited accuracy ISO/IEC 11172-3 audio decoder as defined by the
  standard.

  MAD can alternatively be configured to produce output with less or more
  accuracy than the default, as a tradeoff with performance.

  MAD produces output samples with a precision greater than 24 bits. Because
  most output formats use fewer bits, typically 16, `madplay' implements a
  dithering algorithm when truncating samples for output. This produces high
  quality audio that generally sounds superior to the output of a simple
  rounding algorithm. However, dithering may unfavorably affect an analytic
  examination of the output (such as compliance testing), and therefore it
  may optionally be disabled at runtime.

  It is recommended that applications using the MAD library also implement
  some form of dithering to obtain the best quality audio.

Portability Issues

  GCC is preferred to compile the code, but other compilers may also work.
  The assembly code in `libmad/fixed.h' depends on the inline assembly
  features of your compiler. If you're not using GCC or MSVC++, you can
  either write your own assembly macros or use the default (low quality
  output) version.

  The union initialization of `libmad/huffman.c' may not be portable to all
  platforms when GCC is not used.

  The code should not be sensitive to word sizes or byte ordering, however
  it does assume A % B has the same sign as A.

===============================================================================

BUILDING AND INSTALLING

Windows Platforms

  The MAD library can be built under Windows using either MSVC++ or Cygwin.
  A MSVC++ project file can be found under `msvc++/libmad'. In addition, a
  demo project (based on madplay and complete with ID3 support) can be found
  under `msvc++/demo'.

  To build MAD or the MAD plug-in for Winamp using Cygwin, you will first
  need to install the Cygwin tools:

      http://www.cygwin.com/

  You may then proceed with the following POSIX instructions within the
  Cygwin shell.

POSIX Platforms (including Cygwin)

  The code is distributed with a `configure' script which will generate for
  you a `Makefile' and a `config.h' in each of the various subdirectories
  which require them. See the file `INSTALL' for generic instructions.

  The specific options you may want to give `configure' are:

      --enable-speed            optimize for speed over accuracy

      --enable-accuracy         optimize for accuracy over speed

      --disable-debugging       do not compile with debugging support, and
                                use more optimizations

      --enable-shared		build a shared library if possible

  Note that you need not specify one of --enable-speed or --enable-accuracy;
  in its default configuration, MAD is optimized for both. You should only
  use one of these options if you wish to compromise speed or accuracy for
  the other.

  It is not normally necessary to use the following options, but you may
  fine-tune the configuration with them if desired:

      --enable-fpm=<arch>       use the <arch>-specific version of the
                                fixed-point math assembly routines
                                (current options are: intel, arm, mips,
                                sparc, ppc; also allowed are: 64bit, approx)

      --enable-sso              use the subband synthesis optimization,
                                with reduced accuracy

      --disable-aso             do not use certain architecture-specific
                                optimizations

  By default an appropriate fixed-point assembly routine will be selected
  for the configured host type, if it can be determined. Thus if you are
  cross-compiling for another architecture, you should be sure either to
  give `configure' a host type argument or to use an explicit --enable-fpm
  option.

  If an appropriate assembly routine cannot be determined, the fast
  approximation version will be used. In this case, use of an alternate
  --enable-fpm is recommended.

Experimenting and Developing

  Further options for `configure' that may be useful to developers and
  experimenters are:

      --enable-debugging        enable diagnostic debugging support and
                                debugging symbols

      --enable-profiling        generate `gprof' profiling code

      --enable-experimental     enable code using the EXPERIMENTAL
                                preprocessor define

===============================================================================

COPYRIGHT AND AUTHOR

  Please read the `COPYRIGHT' file for copyright and warranty information.
  Also, the file `COPYING' contains the full text of the GNU GPL.

  Send inquiries, comments, bug reports, suggestions, patches, etc. to:

      Robert Leslie <rob@mars.org>

  See also the MAD home page on the Web:

      http://www.mars.org/home/rob/proj/mpeg/

===============================================================================



CHANGES

 mad - MPEG audio decoder
 Copyright (C) 2000-2001 Robert Leslie

 $Id: CHANGES,v 1.61 2001/10/23 03:26:12 rob Exp $

===============================================================================

Version 0.14.1 (beta)

  * Updated config.guess and config.sub to latest upstream versions.

  * Enabled libtool versioning rather than release numbering.

  * Renamed `libid3' to `libid3tag' and enabled installation as a separate
    library.

  * Enabled the plug-in for Winamp to use libid3tag.

  * Fixed a NetBSD compatibility issue in audio_sun.

  * Fixed a FreeBSD compatibility issue in audio_oss.

  * Modified the audio_null output module to compute peak amplitude.

  * Changed the peak amplitude computation to avoid evaluating log10(0).

  * Improved the documentation in minimad.c.

  * Several other small fixes.

Version 0.14.0 (beta)

  * Added a new ID3 tag manipulation library (libid3). The required zlib
    support is provided either by the host system or by the included static
    library implementation (libz).

  * Fixed a possible segmentation fault due to mmap() abuse.

  * Added a 64-bit FPM negation operation to improve performance of subband
    synthesis on some platforms.

  * Improved MSVC++ portability and added MSVC++ project files.

  * Added rounding to Layer III requantization for slightly better accuracy.

  * Modified the PCM dithering routine for better dithered output.

  * Added TTY controls and new --tty-control and --no-tty-control options to
    `madplay'. TTY control keys: `P' pause, `S' stop, `Q' quit, `F' forward,
    `B' backward.

  * New time/time-remaining modes with --verbose --tty-control (select by
    pressing `T'.)

Version 0.13.0 (beta)

  * The last complete audio frame should now be properly decoded even when
    it is not followed by other data [fixed in both `madplay' and the
    plug-in for Winamp.]

  * The plug-in for Winamp now reads Xing VBR headers for fast length
    calculation.

  * The plug-in for Winamp should now display visuals correctly for 24- and
    32-bit output.

  * Ancillary data is now properly extracted from Layer III streams.

  * Rewrote the Layer III joint stereo decoding routine to correct a major
    MPEG-2 problem and a minor MPEG-1 problem decoding intensity stereo.

  * Eliminated the dependency on sign-extending right shifts for Layer I and
    Layer II.

  * Renamed `private' field to `private_bits' for better C++ compatibility.

  * Gratuitously renamed `sfreq' field to `samplerate' and
    MAD_ERROR_BADSAMPLEFREQ constant to MAD_ERROR_BADSAMPLERATE.

  * Added `samplerate' and `channels' fields to synth.pcm struct to allow
    these to be different from the decoded frame, and for simpler access.

  * Added new mad_stream_options() and mad_decoder_options() API entries for
    special runtime decoding options.

  * Added new MAD_OPTION_IGNORECRC and MAD_OPTION_HALFSAMPLERATE options.

  * Added new --downsample option to `madplay'.

  * Added new MAD_FLAG_FREEFORMAT indicator flag.

  * Fixed some bugs in the async decoder.

  * Added a new mad_timer_multiply() API routine.

  * Eliminated `+' from asm constraints under Intel for better compatibility
    with some compilers.

  * Fixed a PIC-related problem in libmad/imdct_l_arm.S.

  * Eliminated a static variable to make libmad thread-safe.

Version 0.12.5 (beta)

  * Modified Layer III requantization to occur during Huffman decoding for
    significant performance gains.

  * Optimized short block IMDCT by eliminating redundant calculations.

  * Made several other Layer III performance improvements; added
    ASO_INTERLEAVE1, ASO_INTERLEAVE2, and ASO_ZEROCHECK
    architecture-specific options for best performance on various
    architectures.

  * Optimized synthesis DCT to store result values as soon as they are
    calculated.

Version 0.12.4 (beta)

  * New PowerPC fixed-point assembly courtesy of David Blythe.

  * Reorganized fixed-point assembly routines for easier maintenance and
    better performance.

  * Improved performance of subband synthesis through better indexing and
    fewer local variables.

  * Added alias reduction for the lower two subbands of mixed short blocks,
    per a report of ambiguity with ISO/IEC 11172-3 and for uniformity with
    most other implementations. Also improved alias reduction performance
    using multiply/accumulate.

  * Added --enable-strict-iso option to `configure' to override best
    accepted practices such as the alias reduction for mixed short blocks.

  * Improved performance of Layer III IMDCT by using longer
    multiply/accumulate runs where possible.

Version 0.12.3 (beta)

  * Added MPEG 2.5 support.

  * Added preliminary support for parameterizing the binary point position in
    the fixed-point representation.

  * Added calls to madvise() on supported systems to reduce memory overhead
    when using mmap().

  * Added support for automatic mono/stereo channel selection based on output
    capability.

  * Added a new option to force stereo output of single or dual channel
    streams.

  * Added a new option to amplify/attenuate the output signal by an amount
    expressed in decibels.

  * Output statistics (peak amplitude, clipped samples) are now gathered and
    displayed after decoding.

  * Added multiply/accumulate optimization to the Layer III IMDCT for long
    blocks.

  * Fixed a bug in the handling of Layer III mixed_block_flag.

  * Many Winamp plug-in enhancements.

  * Fixed a problem calling setlocale() when NLS is disabled.

  * Fixed a configure problem when multiple -O CFLAGS are present.

Version 0.12.2 (beta)

  * Rearranged the synthesis polyphase filterbank memory vector for better
    locality of reference, and rewrote mad_synth_frame() to accommodate,
    resulting in improved performance.

  * Discovered a combination of compiler optimization flags that further
    improve performance.

  * Changed some array references in layer3.c to pointer derefs.


frame.c
/*
 * mad - MPEG audio decoder
 * Copyright (C) 2000-2001 Robert Leslie
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * $Id: frame.c,v 1.18 2001/04/10 19:26:41 rob Exp $
 */

# ifdef HAVE_CONFIG_H
#  include "config.h"
# endif

# include "global.h"

# include <stdlib.h>

# include "bit.h"
# include "stream.h"
# include "frame.h"
# include "timer.h"
# include "layer12.h"
# include "layer3.h"

static
unsigned long const bitrate_table[5][15] = {
  /* MPEG-1 */
  { 0,  32000,  64000,  96000, 128000, 160000, 192000, 224000,  /* Layer I   */
       256000, 288000, 320000, 352000, 384000, 416000, 448000 },
  { 0,  32000,  48000,  56000,  64000,  80000,  96000, 112000,  /* Layer II  */
       128000, 160000, 192000, 224000, 256000, 320000, 384000 },
  { 0,  32000,  40000,  48000,  56000,  64000,  80000,  96000,  /* Layer III */
       112000, 128000, 160000, 192000, 224000, 256000, 320000 },

  /* MPEG-2 LSF */
  { 0,  32000,  48000,  56000,  64000,  80000,  96000, 112000,  /* Layer I   */
       128000, 144000, 160000, 176000, 192000, 224000, 256000 },
  { 0,   8000,  16000,  24000,  32000,  40000,  48000,  56000,  /* Layers    */
        64000,  80000,  96000, 112000, 128000, 144000, 160000 } /* II & III  */
};

static
unsigned int const samplerate_table[3] = { 44100, 48000, 32000 };

static
int (*const decoder_table[3])(struct mad_stream *, struct mad_frame *) = {
  mad_layer_I,
  mad_layer_II,
  mad_layer_III
};

/*
 * NAME:	header->init()
 * DESCRIPTION:	initialize header struct
 */
void mad_header_init(struct mad_header *header)
{
  header->layer          = 0;
  header->mode           = 0;
  header->mode_extension = 0;
  header->emphasis       = 0;

  header->bitrate        = 0;
  header->samplerate     = 0;

  header->crc_check      = 0;
  header->crc_target     = 0;

  header->flags          = 0;
  header->private_bits   = 0;

  header->duration       = mad_timer_zero;
}

/*
 * NAME:	frame->init()
 * DESCRIPTION:	initialize frame struct
 */
void mad_frame_init(struct mad_frame *frame)
{
  mad_header_init(&frame->header);

  frame->options = 0;

  frame->overlap = 0;
  mad_frame_mute(frame);
}

/*
 * NAME:	frame->finish()
 * DESCRIPTION:	deallocate any dynamic memory associated with frame
 */
void mad_frame_finish(struct mad_frame *frame)
{
  mad_header_finish(&frame->header);

  if (frame->overlap) {
    free(frame->overlap);
    frame->overlap = 0;
  }
}

/*
 * NAME:	decode_header()
 * DESCRIPTION:	read header data and following CRC word
 */
static
int decode_header(struct mad_header *header, struct mad_stream *stream)
{
  unsigned int index;

  header->flags        = 0;
  header->private_bits = 0;

  /* header() */

  /* syncword */
  mad_bit_skip(&stream->ptr, 11);

  /* MPEG 2.5 indicator (really part of syncword) */
  if (mad_bit_read(&stream->ptr, 1) == 0)
    header->flags |= MAD_FLAG_MPEG_2_5_EXT;

  /* ID */
  if (mad_bit_read(&stream->ptr, 1) == 0)
    header->flags |= MAD_FLAG_LSF_EXT;
  else if (header->flags & MAD_FLAG_MPEG_2_5_EXT) {
    stream->error = MAD_ERROR_LOSTSYNC;
    return -1;
  }

  /* layer */
  header->layer = 4 - mad_bit_read(&stream->ptr, 2);

  if (header->layer == 4) {
    stream->error = MAD_ERROR_BADLAYER;
    return -1;
  }

  /* protection_bit */
  if (mad_bit_read(&stream->ptr, 1) == 0) {
    header->flags    |= MAD_FLAG_PROTECTION;
    header->crc_check = mad_bit_crc(stream->ptr, 16, 0xffff);
  }

  /* bitrate_index */
  index = mad_bit_read(&stream->ptr, 4);

  if (index == 15) {
    stream->error = MAD_ERROR_BADBITRATE;
    return -1;
  }

  if (header->flags & MAD_FLAG_LSF_EXT)
    header->bitrate = bitrate_table[3 + (header->layer >> 1)][index];
  else
    header->bitrate = bitrate_table[header->layer - 1][index];

  /* sampling_frequency */
  index = mad_bit_read(&stream->ptr, 2);

  if (index == 3) {
    stream->error = MAD_ERROR_BADSAMPLERATE;
    return -1;
  }

  header->samplerate = samplerate_table[index];

  if (header->flags & MAD_FLAG_LSF_EXT) {
    header->samplerate /= 2;

    if (header->flags & MAD_FLAG_MPEG_2_5_EXT)
      header->samplerate /= 2;
  }

  /* padding_bit */
  if (mad_bit_read(&stream->ptr, 1))
    header->flags |= MAD_FLAG_PADDING;

  /* private_bit */
  if (mad_bit_read(&stream->ptr, 1))
    header->private_bits |= MAD_PRIVATE_HEADER;

  /* mode */
  header->mode = 3 - mad_bit_read(&stream->ptr, 2);

  /* mode_extension */
  header->mode_extension = mad_bit_read(&stream->ptr, 2);

  /* copyright */
  if (mad_bit_read(&stream->ptr, 1))
    header->flags |= MAD_FLAG_COPYRIGHT;

  /* original/copy */
  if (mad_bit_read(&stream->ptr, 1))
    header->flags |= MAD_FLAG_ORIGINAL;

  /* emphasis */
  header->emphasis = mad_bit_read(&stream->ptr, 2);

  if (header->emphasis == 2) {
    stream->error = MAD_ERROR_BADEMPHASIS;
    return -1;
  }

  /* error_check() */

  /* crc_check */
  if (header->flags & MAD_FLAG_PROTECTION)
    header->crc_target = mad_bit_read(&stream->ptr, 16);

  return 0;
}

/*
 * NAME:	free_bitrate()
 * DESCRIPTION:	attempt to discover the bitstream's free bitrate
 */
static
int free_bitrate(struct mad_stream *stream, struct mad_header const *header)
{
  struct mad_bitptr keep_ptr;
  unsigned long rate = 0;
  unsigned int pad_slot, slots_per_frame;
  unsigned char const *ptr = 0;

  keep_ptr = stream->ptr;

  pad_slot = (header->flags & MAD_FLAG_PADDING) ? 1 : 0;
  slots_per_frame = (header->layer == MAD_LAYER_III &&
		     (header->flags & MAD_FLAG_LSF_EXT)) ? 72 : 144;

  while (mad_stream_sync(stream) == 0) {
    struct mad_stream peek_stream;
    struct mad_header peek_header;

    peek_stream = *stream;
    peek_header = *header;

    if (decode_header(&peek_header, &peek_stream) == 0 &&
	peek_header.layer == header->layer &&
	peek_header.samplerate == header->samplerate) {
      unsigned int N;

      ptr = mad_bit_nextbyte(&stream->ptr);

      N = ptr - stream->this_frame;

      if (header->layer == MAD_LAYER_I) {
	rate = (unsigned long) header->samplerate *
	  (N - 4 * pad_slot + 4) / 48 / 1000;
      }
      else {
	rate = (unsigned long) header->samplerate *
	  (N - pad_slot + 1) / slots_per_frame / 1000;
      }

      if (rate >= 8)
	break;
    }

    mad_bit_skip(&stream->ptr, 8);
  }

  stream->ptr = keep_ptr;

  if (rate < 8 || (header->layer == MAD_LAYER_III && rate > 640)) {
    stream->error = MAD_ERROR_LOSTSYNC;
    return -1;
  }

  stream->freerate = rate * 1000;

# if 0 && defined(DEBUG)
  fprintf(stderr, "free bitrate == %lu\n", stream->freerate);
# endif

  return 0;
}

/*
 * NAME:	header->decode()
 * DESCRIPTION:	read the next frame header from the stream
 */
int mad_header_decode(struct mad_header *header, struct mad_stream *stream)
{
  register unsigned char const *ptr, *end;
  unsigned int pad_slot, N;

  ptr = stream->next_frame;
  end = stream->bufend;

  if (ptr == 0) {
    stream->error = MAD_ERROR_BUFPTR;
    goto fail;
  }

  /* stream skip */
  if (stream->skiplen) {
    if (!stream->sync)
      ptr = stream->this_frame;

    if (end - ptr < stream->skiplen) {
      stream->skiplen   -= end - ptr;
      stream->next_frame = end;

      stream->error = MAD_ERROR_BUFLEN;
      goto fail;
    }

    ptr += stream->skiplen;
    stream->skiplen = 0;

    stream->sync = 1;
  }

 sync:
  /* synchronize */
  if (stream->sync) {
    if (end - ptr < MAD_BUFFER_GUARD) {
      stream->next_frame = ptr;

      stream->error = MAD_ERROR_BUFLEN;
      goto fail;
    }
    else if (!(ptr[0] == 0xff && (ptr[1] & 0xe0) == 0xe0)) {
      /* mark point where frame sync word was expected */
      stream->this_frame = ptr;
      stream->next_frame = ptr + 1;

      stream->error = MAD_ERROR_LOSTSYNC;
      goto fail;
    }
  }
  else {
    mad_bit_init(&stream->ptr, ptr);

    if (mad_stream_sync(stream) == -1) {
      if (end - stream->next_frame >= MAD_BUFFER_GUARD)
	stream->next_frame = end - MAD_BUFFER_GUARD;

      stream->error = MAD_ERROR_BUFLEN;
      goto fail;
    }

    ptr = mad_bit_nextbyte(&stream->ptr);
  }

  /* begin processing */
  stream->this_frame = ptr;
  stream->next_frame = ptr + 1;  /* possibly bogus sync word */

  mad_bit_init(&stream->ptr, stream->this_frame);

  if (decode_header(header, stream) == -1)
    goto fail;

  /* calculate frame duration */
  mad_timer_set(&header->duration, 0,
		32 * MAD_NSBSAMPLES(header), header->samplerate);

  /* calculate free bit rate */
  if (header->bitrate == 0) {
    if ((!stream->sync || !stream->freerate) &&
	free_bitrate(stream, header) == -1)
      goto fail;

    header->bitrate = stream->freerate;
    header->flags  |= MAD_FLAG_FREEFORMAT;
  }

  /* calculate beginning of next frame */
  pad_slot = (header->flags & MAD_FLAG_PADDING) ? 1 : 0;

  if (header->layer == MAD_LAYER_I)
    N = ((12 * header->bitrate / header->samplerate) + pad_slot) * 4;
  else {
    unsigned int slots_per_frame;

    slots_per_frame = (header->layer == MAD_LAYER_III &&
		       (header->flags & MAD_FLAG_LSF_EXT)) ? 72 : 144;

    N = (slots_per_frame * header->bitrate / header->samplerate) + pad_slot;
  }

  /* verify there is enough data left in buffer to decode this frame */
  if (N + MAD_BUFFER_GUARD > end - stream->this_frame) {
    stream->next_frame = stream->this_frame;

    stream->error = MAD_ERROR_BUFLEN;
    goto fail;
  }

  stream->next_frame = stream->this_frame + N;

  if (!stream->sync) {
    /* check that a valid frame header follows this frame */

    ptr = stream->next_frame;
    if (!(ptr[0] == 0xff && (ptr[1] & 0xe0) == 0xe0)) {
      ptr = stream->next_frame = stream->this_frame + 1;
      goto sync;
    }

    stream->sync = 1;
  }

  header->flags |= MAD_FLAG_INCOMPLETE;

  return 0;

 fail:
  stream->sync = 0;

  return -1;
}

/*
 * NAME:	frame->decode()
 * DESCRIPTION:	decode a single frame from a bitstream
 */
int mad_frame_decode(struct mad_frame *frame, struct mad_stream *stream)
{
  frame->options = stream->options;

  /* header() */
  /* error_check() */

  if (!(frame->header.flags & MAD_FLAG_INCOMPLETE) &&
      mad_header_decode(&frame->header, stream) == -1)
    goto fail;

  /* audio_data() */

  frame->header.flags &= ~MAD_FLAG_INCOMPLETE;

  if (decoder_table[frame->header.layer - 1](stream, frame) == -1) {
    if (!MAD_RECOVERABLE(stream->error))
      stream->next_frame = stream->this_frame;

    goto fail;
  }

  /* ancillary_data() */

  if (frame->header.layer != MAD_LAYER_III) {
    struct mad_bitptr next_frame;

    mad_bit_init(&next_frame, stream->next_frame);

    stream->anc_ptr    = stream->ptr;
    stream->anc_bitlen = mad_bit_length(&stream->ptr, &next_frame);

    mad_bit_finish(&next_frame);
  }

  return 0;

 fail:
  stream->anc_bitlen = 0;
  return -1;
}

/*
 * NAME:	frame->mute()
 * DESCRIPTION:	zero all subband values so the frame becomes silent
 */
void mad_frame_mute(struct mad_frame *frame)
{
  unsigned int s, sb;

  for (s = 0; s < 36; ++s) {
    for (sb = 0; sb < 32; ++sb) {
      frame->sbsample[0][s][sb] =
      frame->sbsample[1][s][sb] = 0;
    }
  }

  if (frame->overlap) {
    for (s = 0; s < 18; ++s) {
      for (sb = 0; sb < 32; ++sb) {
	(*frame->overlap)[0][sb][s] =
	(*frame->overlap)[1][sb][s] = 0;
      }
    }
  }
}
